Firestore/core/src/firebase/firestore/remote/stream.mm

```diff
@@ -0,0 +1,318 @@
+              "Can only cancel backoff in a stopped state (was %s)", state_);
+
+  // Clear the error condition.
+  state_ = State::Initial;
+  ResetBackoff();
+}
+
+void Stream::ResetBackoff() {
+  EnsureOnQueue();
+  backoff_.Reset();
+}
+
+// Idleness
+
+void Stream::MarkIdle() {
+  EnsureOnQueue();
+
+  if (IsOpen() && !idleness_timer_) {
+    idleness_timer_ = worker_queue_->EnqueueAfterDelay(
+        kIdleTimeout, idle_timer_id_, [this] { Stop(); });
+  }
+}
+
+void Stream::CancelIdleCheck() {
+  EnsureOnQueue();
+  idleness_timer_.Cancel();
+}
+
+// Read/write
+
+void Stream::OnStreamRead(const grpc::ByteBuffer& message) {
+  EnsureOnQueue();
+
+  HARD_ASSERT(IsStarted(), "OnStreamRead called for a stopped stream.");
+
+  if (bridge::IsLoggingEnabled()) {
+    LOG_DEBUG("%s headers (whitelisted): %s", GetDebugDescription(),
+              Datastore::GetWhitelistedHeadersAsString(
+                  grpc_stream_->GetResponseHeaders()));
+  }
+
+  Status read_status = DoOnStreamRead(message);
+  if (!read_status.ok()) {
+    grpc_stream_->Finish();
+    // Don't expect GRPC to produce status -- since the error happened on the
+    // client, we have all the information we need.
+    OnStreamError(read_status);
+    return;
+  }
+}
+
+// Stopping
+
+void Stream::Stop() {
+  EnsureOnQueue();
+
+  LOG_DEBUG("%s stop", GetDebugDescription());
+
+  if (!IsStarted()) {
+    return;
+  }
+
+  RaiseGeneration();
+
+  // If the stream is in the auth stage, GRPC stream might not have been created
+  // yet.
+  if (grpc_stream_) {
+    LOG_DEBUG("%s Finishing GRPC stream", GetDebugDescription());
+    FinishGrpcStream(grpc_stream_.get());
+    ResetGrpcStream();
+  }
+
+  // If this is an intentional close ensure we don't delay our next connection
+  // attempt.
+  ResetBackoff();
+
+  // State must be updated before calling the delegate.
+  state_ = State::Initial;
+  // Stopping the stream was initiated by the client, so we have all the
+  // information we need.
+  DoOnStreamFinish(Status::OK());
+}
+
+void Stream::OnStreamError(const Status& status) {
+  EnsureOnQueue();
+
+  // TODO(varconst): log error here?
+  LOG_DEBUG("%s Stream error", GetDebugDescription());
+
+  RaiseGeneration();
+
+  if (status.code() == FirestoreErrorCode::ResourceExhausted) {
+    LOG_DEBUG(
+        "%s Using maximum backoff delay to prevent overloading the backend.",
+        GetDebugDescription());
+    backoff_.ResetToMax();
+  } else if (status.code() == FirestoreErrorCode::Unauthenticated) {
+    // "unauthenticated" error means the token was rejected. Try force
+    // refreshing it in case it just expired.
+    credentials_provider_->InvalidateToken();
+  }
+
+  ResetGrpcStream();
+
+  // State must be updated before calling the delegate.
+  state_ = State::Error;
+  DoOnStreamFinish(status);
+}
+
+void Stream::ResetGrpcStream() {
+  EnsureOnQueue();
+
+  grpc_stream_.reset();
+  backoff_.Cancel();
+}
+
+// Check state
+
+bool Stream::IsOpen() const {
```

Comment:
These are at the top of the other implementations (before start)

Reply:
Done.